bchrpc/server.go

```diff
@@ -1816,6 +2316,234 @@ func (s *GrpcServer) fetchTransactionsByAddress(addr bchutil.Address, startHeigh
+		isMintBaton    bool = false
+		slpVersionType pb.SlpVersionType
+	)
+
+	slpMsg, _ := v1parser.ParseSLP(entry.SlpOpReturn)
+
+	// set isMintBaton and also check that vout is within proper range
+	if slpMsg.TransactionType == "MINT" {
+		if slpMsg.Data.(v1parser.SlpMint).MintBatonVout == int(vout) {
+			isMintBaton = true
+		} else if vout != 1 {
+			return nil, errors.New("vout is out of range for slp mint")
+		}
+		if slpMsg.TokenType == 0x01 {
+			slpVersionType = pb.SlpVersionType_SLP_V1_MINT
+		} else if slpMsg.TokenType == 0x81 {
+			slpVersionType = pb.SlpVersionType_SLP_NFT1_GROUP_MINT
+		}
+	} else if slpMsg.TransactionType == "GENESIS" {
+		if slpMsg.Data.(v1parser.SlpGenesis).MintBatonVout == int(vout) {
+			isMintBaton = true
+		} else if vout != 1 {
+			return nil, errors.New("vout is out of range for slp genesis")
+		}
+		if slpMsg.TokenType == 0x01 {
+			slpVersionType = pb.SlpVersionType_SLP_V1_GENESIS
+		} else if slpMsg.TokenType == 0x41 {
+			slpVersionType = pb.SlpVersionType_SLP_NFT1_UNIQUE_CHILD_GENESIS
+		} else if slpMsg.TokenType == 0x81 {
+			slpVersionType = pb.SlpVersionType_SLP_NFT1_GROUP_GENESIS
+		}
+	} else if slpMsg.TransactionType == "SEND" {
+		if int(vout) > len(slpMsg.Data.(v1parser.SlpSend).Amounts) {
+			return nil, errors.New("vout is out of range for slp send transaction")
+		}
+		if slpMsg.TokenType == 0x01 {
+			slpVersionType = pb.SlpVersionType_SLP_V1_SEND
+		} else if slpMsg.TokenType == 0x41 {
+			slpVersionType = pb.SlpVersionType_SLP_NFT1_UNIQUE_CHILD_SEND
+		} else if slpMsg.TokenType == 0x81 {
+			slpVersionType = pb.SlpVersionType_SLP_NFT1_GROUP_SEND
+		}
+	}
+
+	// Get decimal amount from Genesis for convenience
+	var decimals int
+	if slpMsg.TransactionType == "GENESIS" {
+		decimals = slpMsg.Data.(v1parser.SlpGenesis).Decimals
+	} else {
+		var _hash []byte
+		for i := len(entry.TokenIDHash) - 1; i >= 0; i-- {
+			_hash = append(_hash, entry.TokenIDHash[i])
+		}
+		_tokenIDRev, _ := chainhash.NewHash(_hash)
+		genEntry, _ := s.getSlpIndexEntry(_tokenIDRev)
+		genSlpMsg, _ := v1parser.ParseSLP(genEntry.SlpOpReturn)
+		decimals = genSlpMsg.Data.(v1parser.SlpGenesis).Decimals
+	}
+
+	// get amount
+	amount, _ := slpMsg.GetVoutAmount(int(vout))
+
+	slpToken := &pb.SlpToken{
+		TokenId:     entry.TokenIDHash[:],
+		Amount:      amount.Uint64(),
+		IsMintBaton: isMintBaton,
+		Decimals:    uint32(decimals),
+		VersionType: slpVersionType,
+	}
+
+	return slpToken, nil
+}
+
+// manageSlpEntryCache keeps the SlpEntryCache updated on transaction and block events
+//
+// TODO: need to remove old items from this map with a timer
+//
+func (s *GrpcServer) manageSlpEntryCache() {
+
+	if s.slpIndex == nil {
+		return
+	}
+
+	subscription := s.subscribeEvents()
+	defer subscription.Unsubscribe()
+
+	for {
+		event := <-subscription.Events()
+		switch event := event.(type) {
+		case *rpcEventTxAccepted:
+			txDesc := event
+
+			err := s.slpIndex.AddMempoolTx(txDesc.Tx)
+			if err != nil {
+				continue
```

Comment:
We should log failures here for tracking.

Reply:
done.